Add Vararg support in embedded defs

fschueler opened the pull request:
This is my first stab at adding vararg support in embedded defs.

My approach is to change the repeated parameters into a `Seq` for lambda application and then wrap the original parameters into `Seq.apply()` in the Lowering.

This does what I want in my usecase but I can't figure out how to write a test for it that does what I want (see inline note).

Also I am not sure about the last missing case for VarargsSpliced - what is the meaning of this?

LPTK commented on src/main/scala/squid/ir/Lowering.scala:
```diff
@@ -41,7 +41,12 @@ trait Lowering extends Transformer {
+              val ruh = RuntimeUniverseHelpers
+              val functionArity = reps.size + vreps.size
+              val typ = r.typ.typeArgs.last
+              val varargsAsSeq = base.rep(MethodApp(base.rep(StaticModule("scala.collection.Seq")), ruh.FunctionType.symbol(vreps.size).toType member ruh.sru.TermName("apply") asMethod, Nil, Args(vreps:_*)::Nil, ruh.SequenceType.apply(typ)))
```
It looks like you're calling `FunctionN`'s `apply` method on object `Seq` here! It's wrong, as `Seq` has its own `apply` method (inherited from `scala.collection.generic.GenericCompanion `).
There is actually no need for messing with runtime universes here. You can see [in this code](https://github.com/epfldata/squid/blob/master/src/main/scala/squid/ir/AST.scala#L214-L221) a working example of building a `Seq` from a list of arguments ;^)

LPTK commented:
Very nice! The functionality has been on my backburner forever – like so many other things, unfortunately – so it's cool to see it getting implemented.

Note: like last time, I'm currently at a conference preparing a talk, so I may not be very reactive in the following days!

LPTK commented:
> Also I am not sure about the last missing case for VarargsSpliced - what is the meaning of this?

The meaning of `ArgsVarargSpliced` is explained [here](https://github.com/epfldata/squid/blob/master/core/src/main/scala/squid/lang/Base.scala#L132).

fschueler commented:
Haha, I thought you must be at POPL right now ;)

Thanks for commenting though, that's very helpful! I will change the `Seq` stuff and try to include the `ArgsVarargSpliced` case!

fschueler commented:
I've changed the `Seq.apply` construction and added a case for `ArgsVarargSpliced` but can't really figure out how to write tests that do what I want in `ClassEmbeddingTests`. It seems like there is a fundamental thing about the code"" quasiquotes that I haven't fully understood yet. 
I mainly can't grasp how the right hand side of the `code"code under test" eqt code"expected code"` call gets expanded.

When you have a minute maybe we can chat over at gitter, for now good luck with your POPL talk! :)

LPTK commented on core/src/main/scala/squid/quasi/ClassEmbedding.scala:
```diff
@@ -182,7 +182,13 @@ class ClassEmbedding(override val c: whitebox.Context) extends QuasiMacros(c) {
+        ValDef(vd.mods, vd.name, newTypeTree, EmptyTree)
+
+      case _ => ValDef(vd.mods, vd.name, vd.tpt, EmptyTree)
+    } // TODO handle by-name and repeated
```
Can you move the comment back to the top of the `def` and make it `TODO handle by-name arguments`?

LPTK commented on core/src/main/scala/squid/utils/meta/UniverseHelpers.scala:
```diff
@@ -99,7 +99,16 @@ trait UniverseHelpers[U <: scala.reflect.api.Universe] {
   }
-  
+
+  object SequenceType {
```
This object seems to be unused

fschueler replied:
A relic from my first tries...removed!

LPTK commented on src/main/scala/squid/ir/Lowering.scala:
```diff
@@ -41,7 +41,29 @@ trait Lowering extends Transformer {
-            case (r, args) => lastWords(s"Not supported yet: vararg $args") // TODO
+            case (r, ArgsVarargs(Args(reps @ _*), Args(vreps @ _*))) =>
+              val ruh = RuntimeUniverseHelpers
+              val functionArity = reps.size + vreps.size
```
`vreps.size` is wrong, as all the varargs as packed as one argument of type `Seq[typ]`

LPTK commented on src/main/scala/squid/ir/Lowering.scala:
```diff
@@ -41,7 +41,29 @@ trait Lowering extends Transformer {
+            case (r, ArgsVarargs(Args(reps @ _*), Args(vreps @ _*))) =>
+              val ruh = RuntimeUniverseHelpers
+              val functionArity = reps.size + vreps.size
+              val typ = r.typ.typeArgs.last
```
Why would `typ` in the `Seq[typ]` type of the vararg argument correspond to the last type argument to the method!? Very wild assumption here 😛

fschueler replied:
Hah, good catch, this is definitely the wrong type. I think I just copied it from the regular case above.

But now I'm confused: I saw in your PR that you changed the construction of the `typ` to use the least upper bound for the typ of the sequence constructor and that makes sense to me. 

But it's also used for the type of the final `methodApp` construction. Shouldn't that be a different type here? As far as I understood `r.typ.typeArgs.last` gives us the return type of the `Def`, right?

In that case I would change the code so that it uses the `lub` of the `vreps` for the `Seq` type and the `r.typ.typeArgs.last` as the last argument to `methodApp`.

LPTK replied:
You're right, good catch! I made the modification too fast and forgot that `typ` was also used later on (in that humongous `MethodApp` line that I should I broken in two distinct lines anyway).

Also, relative to using `r.typ.typeArgs.last`: it took me some time to remember (I should have _definitely_ left a comment about such a non-trivial mechanism), but the actual reason such a type works as the method's return type is because at this point `r` has to be a function type, whose last type argument indicates its return type (so my own earlier "wild assumption" comment about that type was moot).

That's what happens when you come back to code you haven't touched in many months 😅

fschueler replied:
Thanks, that makes sense. I added the comment to my commit ;)

LPTK commented on src/test/scala/squid/classembed/ClassEmbeddingTests.scala:
```diff
@@ -33,40 +33,64 @@ class ClassEmbeddingTests extends MyFunSuite {
   test("Basics") {
     
     Emb.Object.Defs.foo.value eqt code"(arg: Int) => MyClass.foo(arg.toLong).toInt"
-    
```
Please configure your editor _not_ to remove indentation spaces, and rebase the commits to remove these changes.

LPTK commented on src/test/scala/squid/classembed/ClassEmbeddingTests.scala:
```diff
@@ -33,40 +33,64 @@ class ClassEmbeddingTests extends MyFunSuite {
+
+    OrphanObject.embedIn(TestDSL).Object.Defs.varargFoo.value eqt code"(args: Seq[Int]) => args.sum + 1"
+
+    OrphanObject.embedIn(TestDSL).Object.Defs.varargBar[Int] eqt code"(args: Seq[Int]) => args.length"
```
This does not compile (there is no such `varargBar` defined).

fschueler commented:
Thanks for taking the time to review the PR!

I addressed your changes and hope nothing else is wrong or missing now. My tests pass and do what I expect them to do ;)

LPTK commented:
Great contribution, thanks!
